Add iter_entities to World (#6228)

ramirezmike opened the pull request:
# Objective

- Add a way to iterate over all entities from &World

## Solution

- Added a function `iter_entities` on World which returns an iterator of `Entity` derived from the entities in the `World`'s `archetypes`

---

## Changelog

- Added a function `iter_entities` on World, allowing iterating over all entities in contexts where you only have read-only access to the World.

harudagondi commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -320,6 +320,15 @@ impl World {
+    pub fn iter_entities(&self) -> impl Iterator<Item = Entity> + '_ {
+        self.archetypes
+            .iter()
+            .flat_map(|archetype| archetype.entities().iter().cloned())
```
```suggestion
            .flat_map(|archetype| archetype.entities().iter().copied())
```

ramirezmike replied:
Updated to use copied. Thanks!

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -320,6 +320,15 @@ impl World {
     }
 
+    /// Returns an [`Entity`](crate::entity::Entity) iterator of current entities.
+    /// This is useful in contexts where you only have read-only access to the [World].
```
```suggestion
    ///
    /// This is useful in contexts where you only have read-only access to the [`World`].
```

Semantic white space. This affects the tooltips on IDEs; we deliberately separate out the short description on the first line from helpful details.

You also missed the backticks :)

ramirezmike replied:
Ooph, sorry about that; good to know for the future.

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -320,6 +320,15 @@ impl World {
         Some(EntityRef::new(self, entity, location))
     }
 
+    /// Returns an [`Entity`](crate::entity::Entity) iterator of current entities.
```
```suggestion
    /// Returns an [`Entity`] iterator of current entities.
```

This link can be collapsed, as `Entity` is already in scope.

ramirezmike replied:
Updated in latest change. Thanks!

mockersf commented:
Could you use this new function in the scene code from #6227?

alice-i-cecile commented:
Yep, it would be ideal to clean that code up while we're here. If you don't have time, I'll merge this on Monday and make a new easy.

ramirezmike commented:
> Could you use this new function in the scene code from #6227?

I rebased and added a commit to update dynamic_scene.rs to use iter_entities. I know the code is pretty much the same and all the tests pass, but I'll admit, I'm not _too_ familiar with the behavior of dynamic_scene. I ran the scene/scene example before and after the changes and compared the output and it was the same.

Let me know if you'd like me to do more extensive testing though 👍
